src/DynamicNLPModels.jl

```diff
@@ -366,14 +367,28 @@ function _build_sparse_lq_dynamic_model(dnlp::LQDynamicData{T, V, M, MK}) where
     gl = dnlp.gl
     gu = dnlp.gu
 
-    H   = _build_H(Q, R, N; Qf = Qf, S = S)
-    J1  = _build_sparse_J1(A, B, N)
-    J2  = _build_sparse_J2(E, F, N)
+    nc = size(E, 1)
 
-    J = vcat(J1, J2)
+    H_colptr = _init_similar(Int[], ns * (N + 1) + nu * N + 1, Int)
+    H_rowval = _init_similar(Int[], (ns + nu) * N * ns + (ns + nu) * N * nu + ns * ns, Int)
+    H_nzval  = _init_similar(T[], (ns + nu) * N * ns + (ns + nu) * N * nu + ns * ns, T)
 
-    c0  = zero(T)
+    J_colptr = _init_similar(Int[], ns * (N + 1) + nu * N + 1, Int)
+    J_rowval = _init_similar(Int[], N * (ns^2 + ns * nu + ns) + N * (nc * ns + nc * nu), Int)
+    J_nzval  = _init_similar(T[],   N * (ns^2 + ns * nu + ns) + N * (nc * ns + nc * nu), T)
+
+    _set_sparse_H!(H_colptr, H_rowval, H_nzval, Q, R, N; Qf = Qf, S = S)
+
+    H = SparseArrays.SparseMatrixCSC((N + 1) * ns + nu * N, (N + 1) * ns + nu * N, H_colptr, H_rowval, H_nzval)
 
+    _set_sparse_J!(J_colptr, J_rowval, J_nzval, A, B, E, F, K, N)
+
+    J = SparseArrays.SparseMatrixCSC((nc + ns) * N, (N + 1) * ns + nu * N, J_colptr, J_rowval, J_nzval)
+
+    SparseArrays.dropzeros!(H)
+    SparseArrays.dropzeros!(J)
```

Comment:
Can we go even further? Can we preallocate the memory based on the sparsity of `Q,R,S`, and remove the `dropzeros!` step?

Reply:
But if this is not obvious & the current implementation is efficient enough, I suggest opening an issue and improving it later